### PR TITLE
Add option to pass a custom root element to IntersectionObserver

### PR DESCRIPTION
**Pull request by sreedhar7**

### CODE REVIEW

File: src/metal/scheduler.ts [hunk 1]
```diff
@@ -25,18 +25,38 @@ let tokenCounter = 0;
...
-    public scrollTop: number,
     public scrollLeft: number,
+    public scrollTop: number,
+    public x: number,
+    public y: number,
     public width: number,
     public height: number
   ) {}
-  static generate(): Frame {
+  static generate(root?: Element): Frame {
```
Comment (asakusuma) [inline src/metal/scheduler.ts]: We probably want to have a different function for generating the Frame with a root. Having a significant branch like this on a hot code path is probably not idea. Thoughts @krisselden @chadhietala?

File: test/app/setup.js [hunk 2]
```diff
@@ -15,3 +15,13 @@ for (var i = 0;  i < 100; i++) {
   app.appendChild(el);
 }
 
+var root = document.getElementById('root');
+for (var i = 0;  i < 5; i++) {
```
Comment (asakusuma) [inline test/app/setup.js]: We should have more elements so you can scroll the 3rd element completely out of the root viewport

File: test/headless/specs/intersection-observer.js [hunk 3]
```diff
@@ -230,4 +230,34 @@ testModule('IntersectionObserver', class extends TestClass {
...
+      root.scrollTop = 100;
+    })
+    .wait(100)
+    .evaluate(function() {
+      root.scrollTop = 10;
+    })
+    .wait(100)
+    .getExecution()
+    .evaluate(function() {
+      return window.STATE.impressions;
```
Comment (asakusuma) [inline test/headless/specs/intersection-observer.js]: Can we also assert that the `intersectionRatio` is a reasonable number and that the `rootBounds` is correct?

File: src/metal/scheduler.ts (hunk 1 above)
Comment (asakusuma) [inline src/metal/scheduler.ts]: This hotpath is already not great, needs work related to https://github.com/linkedin/spaniel/issues/12

**Comment by asakusuma** [comment]
This will address https://github.com/linkedin/spaniel/issues/50

**Comment by lynchbomb** [comment]
FWIW: This functionality is currently at the top of API feature requests from folks.

**Comment by sreedhar7** [comment]
@lynchbomb @asakusuma Sorry I don't have enough time to work on this for the next couple of months. Please feel free to improve my changes and merge into master.

**Comment by asakusuma** [comment]
Already implemented by https://github.com/linkedin/spaniel/pull/53

**Comment by asakusuma** [comment]
Closing

**Pull request closed without merging**